2.0 Upgrade guide (#3467)
DataDog/dd-trace-rb: docs/UpgradeGuide2.md

# Upgrading ddtrace to 2.0

## From 1.x to 2.0

Upgrading `ddtrace` from 1.x to 2.0 introduces some breaking changes which are outlined below.

[**Upgrade basic usage**](#2.0-upgrading-basic-usage)

For users with a basic implementation (configuration file + out-of-the-box instrumentation), only minor changes to your configuration file are required: most applications take just minutes to update. Check out the following sections for a step-by-step guide.

- [Requires Ruby 2.5+](#2.0-requires-ruby-2.5+)
- [Extracts datadog-ci gem](#2.0-extracts-ci)
- [Configuration changes](#2.0-configuration-changes)
  - [Replace `use` with `instrument`](#2.0-use-instrument)
  - [Type checking](#2.0-type-checking)
  - [Propagation default](#2.0-propagation-default)
  - [Options](#2.0-options)

---

[**Upgrade advanced usage**](#2.0-upgrading-advanced-usage)

> For users with an advanced implementation of `ddtrace` (custom instrumentation, sampling or processing behavior, etc), some additional may be required. See the following sections for details about what features changed and how to use them.

Reviewer: I believe a word is missing here like `steps` or `changes` or `configuration changes`?

> - [Frozen String](#2.0-frozen-string)

Reviewer: nit: Would `Frozen String Literal` be more descriptive?

- [Tracing API](#2.0-tracing-api)
- [Configuration API](#2.0-configuration-api)
- [Log Correlation](#2.0-log-correlation)
- [Distributed Tracing](#2.0-distributed-tracing)
- [Transport](#2.0-transport)
- [Sampling](#2.0-sampling)

---

> [**Upgrade Instrumentations**](#2.0-upgrade-instrumentations)

Reviewer: We use the word upgrade in several of the headlines like `Upgrade advanced usage` and `Upgrade basic usage`. Is that a bit redundant since this is the upgrade guide? Should we just drop the word `upgrade` from these sections?

> Outlines the changes for our instrumentations (ie: change of support policy and option behaviour).

Reviewer: It's not merged yet, but it's safe to assume the support policy can be [linked here](https://github.com/DataDog/dd-trace-rb/blob/a4ee62e46947cb91cb53eb7beba22bf070d22101/docs/Compatibility.md#support-policy).

If you'd rather wait for that to be merged, or for me to update this upgrade guide if it gets merged first that's fine.

Author: >It's not merged yet, but it's safe to assume the support policy can be [linked here](https://github.com/DataDog/dd-trace-rb/blob/a4ee62e46947cb91cb53eb7beba22bf070d22101/docs/Compatibility.md#support-policy). (remove the branch blob id, of course)

I think mentioning support policy changes here is confusing, I remove it.

- [Error Handling](#2.0-error-handling)
- [Option `error_status_codes`](#2.0-error-status-codes)
- [List of Integration changes](#2.0-list-of-integration-changes)

<h2 id="2.0-upgrading-basic-usage">Upgrading Basic Usage</h2>

<h3 id="2.0-requires-ruby-2.5+">Requires Ruby 2.5+</h3>

The minimum Ruby version requirement for ddtrace 2.x is 2.5.0. For prior Ruby versions, you can use ddtrace 1.x. see more with our support policy.

<h3 id="2.0-extracts-ci">Extracts datadog-ci gem</h3>

> CI visibility component has been extracted as a separate gem named [datadog-ci](https://github.com/DataDog/datadog-ci-rb), and will no longer be installed together with ddtrace.

Reviewer: ```suggestion
The CI visibility component has been extracted as a separate gem named [datadog-ci](https://github.com/DataDog/datadog-ci-rb), and will no longer be installed together with ddtrace.
```

If you are using our CI visibility product, include `datadog-ci` in your Gemfile and learn more about the [setup](https://github.com/DataDog/datadog-ci-rb).

```ruby
gem 'ddtrace', '>= 2'
gem 'datadog-ci'
```

> If you do not want to install `datadog-ci`, make sure to remove CI-related configuration( `Datadog.configure { |c| c.ci.* }`)

Author: I held the opposite opinion on this.

> If you do not want to install `datadog-ci`, make sure to remove CI-related configuration( `Datadog.configure { |c| c.ci.* }`)

Reviewer 2: Sorry, I didn't quite get what you mean by "held the opposite opinion"?

<h3 id="2.0-configuration-changes">Configuration changes</h3>

<h4 id="2.0-use-instrument">Replace `use` with `instrument`</h4>

```ruby
# === with 1.x ===
Datadog.configure do |c|
  c.tracing.use :mysql2
end

# === with 2.0 ===
Datadog.configure do |c|
  c.tracing.instrument :mysql2
end
```

<h4 id="2.0-type-checking">Enforce type checking</h4>

> Configuration options are type checked. When validation fails, it raises `ArgumentError`.

Reviewer: ```suggestion
Configuration options are type checked. When validation fails, an `ArgumentError` is raised.
```

For example `c.env` and `c.service` now have to be `String`.

```ruby
# === with 1.x ===
Datadog.configure do |c|
  c.env = :production
end

# === with 2.0 ===
Datadog.configure do |c|
  c.env = 'production'
end
```

> Notes: Skipping validation with `ENV['DD_EXPERIMENTAL_SKIP_CONFIGURATION_VALIDATION']` has been removed.

Reviewer 2: ```suggestion
Note that the `DD_EXPERIMENTAL_SKIP_CONFIGURATION_VALIDATION` option from ddtrace 1.x has also been removed.
```

> <!-- Here's the list of all affected options:

Reviewer: probably want to remove this comment

  | `diagnostics.health_metrics.enabled`                    | `health_metrics.enabled`            | `DD_HEALTH_METRICS_ENABLED`          |
  | `diagnostics.health_metrics.statsd`                     | `health_metrics.statsd`             | (none)                               |
  | `profiling.advanced.max_events`                         | Removed                             | (none)                               |
  | `profiling.advanced.legacy_transport_enabled`           | Removed                             | (none)                               |
  | `profiling.advanced.force_enable_new_profiler`          | Removed                             | (none)                               |
  | `profiling.advanced.force_enable_legacy_profiler`       | Removed                             | (none)                               |

<h2 id="2.0-upgrading-advanced-usage">Upgrading Advanced Usage</h2>

<h3 id="2.0-frozen-string">Frozen String</h3>

All strings are frozen by default. Make sure your code does not mutate them.

<h3 id="2.0-configuration-api">Configuration API</h3>

If you are [writing your own instrumentation](DevelopmentGuide.md#writing-new-integrations),

> configuration options are now lazily evaluated by default, hence remove `.lazy` from option definition.

Reviewer: ```suggestion
configuration options are now lazily evaluated by default. The `.lazy` option needs to be removed from all option configurations.
```

```ruby
class MySettings < Datadog::Tracing::Contrib::Configuration::Settings
  option :boom do |o|
    o.default do
      true
    end
    o.lazy # === Remove this with 2.0 ===
  end
end
```

<h3 id="2.0-tracing-api">Tracing API</h3>

> Remove option `span_type` from `Datadog::Tracing.trace` method and the following alias methods

Reviewer: ```suggestion
Remove the option `span_type` from the `Datadog::Tracing.trace` method. Additionally, the following alias methods have been removed:
```

| `Datadog::Tracing.trace(name, span_type: ...)` | `Datadog::Tracing.trace(name, type: ...)` |
| `Datadog::Tracing::SpanOperation#span_id`      | `Datadog::Tracing::SpanOperation#id`      |
| `Datadog::Tracing::SpanOperation#span_type`    | `Datadog::Tracing::SpanOperation#type`    |
| `Datadog::Tracing::SpanOperation#span_type=`   | `Datadog::Tracing::SpanOperation#type=`   |
| `Datadog::Tracing::Span#span_id`               | `Datadog::Tracing::Span#id`               |
| `Datadog::Tracing::Span#span_type`             | `Datadog::Tracing::Span#type`             |
| `Datadog::Tracing::Span#span_type=`            | `Datadog::Tracing::Span#type=`            |

If you are using [manual instrumentation](GettingStarted.md#manual-instrumentation) or [processing pipeline](GettingStarted.md#processing-pipeline)

```ruby
# === with 1.x ===
Datadog::Tracing.trace('my_span', span_type: 'custom') do |span|
  puts span.span_id

  span.span_type = "...."
end

# === with 2.0 ===
Datadog::Tracing.trace('my_span', type: 'custom') do |span|
  puts span.id

  span.type = "...."
end
```

<h3 id="2.0-log-correlation">Log Correlation</h3>

> Remove obsolete fields from `Datadog::Tracing::Correlation::Identifier`, and responds to `#env`, `#service`, `#version`, `#trace_id`, `#span_id` and `#to_log_format`. The values returned from `#trace_id` and `#span_id` change from `Integer` to `String`.

Reviewer: This paragraph could probably be made clearer. Maybe something more like "The following fields have been removed from `Datadog::Tracing::Correlation::Identifier and it no longer responds to...

I would also maybe separate `The values returned from....` into a separate paragraph.

If you are [manually correlating logs](GettingStarted.md#trace-correlation), check if it is still compatible.

```ruby
# === with 1.x ===
Datadog::Tracing.correlation.span_id
# => 50288418819650436

# === with 2.0 ===
Datadog::Tracing.correlation.span_id
# => '50288418819650436'
```

<h3 id="2.0-distributed-tracing">Distributed Tracing</h3>

<h4 id="2.0-distributed-tracing-api">Propagation API changes</h4>

If you are [manually propagating distributed tracing metadata](GettingStarted.md#using-the-http-propagator) `Datadog::Tracing::Propagation::HTTP` has moved to `Datadog::Tracing::Contrib::HTTP`.

```ruby
# === with 1.x ===
Datadog::Tracing::Propagation::HTTP.inject!
Datadog::Tracing::Propagation::HTTP.extract

# === with 2.0 ===
Datadog::Tracing::Contrib::HTTP.inject
Datadog::Tracing::Contrib::HTTP.extract
```

<h4 id="2.0-distributed-tracing-env">Environment variable changes</h4>

| 1.x                            | 2.0                                  |
| ------------------------------ | ------------------------------------ |
| `DD_PROPAGATION_STYLE_INJECT`  | `DD_TRACE_PROPAGATION_STYLE_INJECT`  |
| `DD_PROPAGATION_STYLE_EXTRACT` | `DD_TRACE_PROPAGATION_STYLE_EXTRACT` |

> 

Reviewer: Perhaps it is my own ignorance on the topic, so feel free to ignore but it isn't clear to me why values becoming case-insensitive means `the values mapped to different b3 strategies also changed`? Since the values are case-insensitive, it isn't necessary to change existing ones?

The values from the environment variables `DD_TRACE_PROPAGATION_STYLE`, `DD_TRACE_PROPAGATION_STYLE_INJECT`, and `DD_TRACE_PROPAGATION_STYLE_EXTRACT` are now considered case-insensitive. Hence, the values mapped to different b3 strategies (single header vs. multiple headers) also changed.

| Constant                                                                                | Value     | Strategy         |
| --------------------------------------------------------------------------------------- | --------- | ---------------- |
| `Datadog::Tracing::Configuration::Ext::Distributed::PROPAGATION_STYLE_B3_SINGLE_HEADER` | `b3`      | single header    |
| `Datadog::Tracing::Configuration::Ext::Distributed::PROPAGATION_STYLE_B3_MULTI_HEADER`  | `b3multi` | multiple headers |

<h4>Constant changes</h4>

> Remove constants at `Datadog::Tracing::Distributed::Headers::Ext`. see table below:

Author: Both.

> Remove constants at `Datadog::Tracing::Distributed::Headers::Ext`. see table below:

Reviewer 2: So I guess this should be clarified in the doc?

| `Datadog::Tracing::Sampling::RateByServiceSampler` | Privatized |
| `Datadog::Tracing::Sampling::RateLimiter`          | Privatized |
| `Datadog::Tracing::Sampling::TokenBucket`          | Privatized |
| `Datadog::Tracing::Sampling::UnlimitedLimiter`     | Privatized |
| `Datadog::Tracing::Sampling::RateSampler`          | Privatized |
| `Datadog::Tracing::Sampling::Rule`                 | Privatized |
| `Datadog::Tracing::Sampling::SimpleRule`           | Privatized |
| `Datadog::Tracing::Sampling::RuleSampler`          | Privatized |
| `Datadog::Tracing::Sampling::Sampler`              | Privatized |

<h2 id="2.0-upgrade-instrumentations">Upgrade Instrumentations</h2>

<h3 id='2.0-error-handling'>Error Handling</h3>

The `error_handler` options have been replaced by `on_error` to align with `options` for our public API `Datadog::Tracing.trace(name, options)`.

For the majority of integrations, rename the `error_handler` option to `on_error` in your configuration. See details for [`active_job`](#activejob), [`grpc`](#grpc), [`faraday`](#faraday) and [`excon`](#excon), which have unique implementation changes.

<h3 id='2.0-error-status-codes'>Option `error_status_codes`</h3>

> Option `error_status_codes` is introduced to various http integrations. It tags the span with an error based on http status from response header. Its value can be a range (`400...600`), or an array of ranges/integers `[403, 500...600]`. If configured with environment variable, use dash for an end-excluded range (`'400-599'`) and comma for adding element into an array (`'403,500-599'`)

Reviewer: ```suggestion
Option `error_status_codes` has been introduced to various http integrations. It tags the span with an error based on http status from a response header. Its value can be a range (`400...600`), or an array of ranges/integers `[403, 500...600]`. If configured with environment variable, use a dash for an end-excluded range (`'400-599'`) and a comma for adding element into an array (`'403,500-599'`)
```

> Option `error_status_codes` is introduced to various http integrations. It tags the span with an error based on http status from response header. Its value can be a range (`400...600`), or an array of ranges/integers `[403, 500...600]`. If configured with environment variable, use dash for an end-excluded range (`'400-599'`) and comma for adding element into an array (`'403,500-599'`)

Reviewer: I wonder if this should include code examples outside of the paragraph?

#### GRPC

- `error_handler`, `server_error_handler` and `client_error_handler` options are removed. Replace them with option `on_error`, which is invoked on both server and client side instrumentation. Merge your implementation for `server_error_handler` and `client_error_handler` to `on_error`. The implementation for `on_error` should distinguish between the server and client.

  ```ruby
  Datadog.configure do |c|
    c.tracing.instrument :grpc, on_error: proc do |span, error|
      if span.name == 'grpc.service'
        # Do something for server instrumentation
      end

      if span.name == 'grpc.client'
        # Do something for client instrumentation
      end
    end
  end
  ```

#### Net/Http

- `Datadog::Tracing::Contrib::HTTP::Instrumentation.after_request` has been removed.

#### OpenTracing

- Removed entirely. Use `ddtrace` 1.x.

#### PG

- Rename `error_handler` option to `on_error`. See [Error Handling](#2.0-error-handling)

#### Qless

- Removed entirely. Use `ddtrace` 1.x.

#### Que

- Rename `error_handler` option to `on_error`. See [Error Handling](#2.0-error-handling)

<h4 id='2.0-instrumentation-rack'>Rack</h4>

> - The type for `request_queuing` option is `Boolean`. When enabled, the behaviour changes from 1 span to 2 spans. The original `http_server.queue` span rename to `http.proxy.request`, with an additional `http.proxy.queue` span representing the time spent in a load balancer queue before reaching application.

Reviewer: `behaviour`

I'm not sure what the actual style guide/policy is on US vs UK English, but this is the UK spelling. The US spelling is `behavior`.

  ```ruby
  # === with 1.x ===
  Datadog.configure do |c|
    c.tracing.instrument :rack, request_queuing: :include_request
    # or
    c.tracing.instrument :rack, request_queuing: :exclude_request
  end

  # === with 2.0 ===
  Datadog.configure do |c|
    c.tracing.instrument :rack, request_queuing: true
  end
  ```

>   Changing the name of the top-level span (`http_server.queue` -> `http.proxy.request`) would break a lot of stuff (monitors, dashboards, notebooks). The following snippet rename the top-level span back to help migration.

Reviewer: `stuff`

I feel like this is too informal a word for a technical doc like this. I think `things` or `features` or `functionality` would be a better choice.

>   Changing the name of the top-level span (`http_server.queue` -> `http.proxy.request`) would break a lot of stuff (monitors, dashboards, notebooks). The following snippet rename the top-level span back to help migration.

Reviewer: ```suggestion
  Changing the name of the top-level span (`http_server.queue` -> `http.proxy.request`) would break functionality such as monitoring, dashboards and notebooks. The following snippet renames the top-level span back to assist with migration.
```

# Upgrading ddtrace from 1.x to 2.0

Upgrading `ddtrace` from 1.x to 2.0 introduces some breaking changes which are outlined below.

[**Basic Usage**](#2.0-basic-usage)

In most cases, (e.g. when you use out-of-the-box instrumentation and a configuration file) only minor changes to your current setup are needed: most applications take just minutes to update.

- [Requires Ruby 2.5+](#2.0-requires-ruby-2.5+)
- [Extracts datadog-ci gem](#2.0-extracts-ci)
- [Configuration changes](#2.0-configuration-changes)
  - [Replace `use` with `instrument`](#2.0-use-instrument)
  - [Type checking](#2.0-type-checking)
  - [Propagation default](#2.0-propagation-default)

>   - [Options](#2.0-options)

Author: I would like to add the datadog rename afterward.

- [Tracing API](#2.0-tracing-api)
- [Configuration API](#2.0-configuration-api)
- [Log Correlation](#2.0-log-correlation)
- [Distributed Tracing](#2.0-distributed-tracing)
- [Transport](#2.0-transport)
- [Sampling](#2.0-sampling)

---

[**Upgrade Instrumentations**](#2.0-upgrade-instrumentations)

Outlines the changes for our instrumentations.

- [Error Handling](#2.0-error-handling)
- [Option `error_status_codes`](#2.0-error-status-codes)
- [List of Integration changes](#2.0-list-of-integration-changes)

<h2 id="2.0-basic-usage">Basic Usage</h2>

<h3 id="2.0-requires-ruby-2.5+">Requires Ruby 2.5+</h3>

The minimum Ruby version requirement for ddtrace 2.x is 2.5.0. For prior Ruby versions, you can use ddtrace 1.x. see more with our support policy.

<h3 id="2.0-extracts-ci">Extracts datadog-ci gem</h3>

The CI visibility component has been extracted as a separate gem named [datadog-ci](https://github.com/DataDog/datadog-ci-rb), and will no longer be installed together with ddtrace.

If you are using our CI visibility product, include `datadog-ci` in your Gemfile and learn more about the [setup](https://github.com/DataDog/datadog-ci-rb).

```ruby
gem 'ddtrace', '>= 2'

> gem 'datadog-ci'

Reviewer 3: minor: I would recommend putting datadog-ci under test group so that it does not load in production:

```ruby
gem 'ddtrace', '>= 2'

group :test do
  gem 'datadog-ci'
end
```
